### PR TITLE
[NEEDS TESTING] Flockmind

### PR DESCRIPTION
**Pull request by ghost**
Another stupid antag pr

# Document the changes in your pull request

It is hard to call this a port, since it doesn't use any code from the original goonstation antag(because it isn't compatible), but it still does use the idea and will use some sprites from the goon flockmind.

Flockmind is a funny antagonist that makes buildings and drones, convert shit, break down into resources and other funny actions. Mostly it acts through flockdrones, that are AI controlled by default - they will try to consume items, convert turfs into flock-versions and shoot down mobs/mechs. They AI will not be so smart as in goon flockmind(because they wouldn't have any special AI since i dont know how to do that), but they would have something like smart-shooting - they have two weapons, a weak disabler and a even more weak lethal laser, and they will use the disabler on stunable opponents, and if their opponent isn't stunable they will try to kill them. Flockdrones will be able to be possesed by flocktraces summoned by their flockmind, and by their flockmind. Flockmind has many abilities to use, flocktraces have less abilities. Their ultimate goal is to summon The Relay and protect it for 6 minutes.

Check the wiki if you want more info, through it will not be a 100% copy: https://wiki.ss13.co/Flockmind

# Spriting

image
image
image
image
image

# Wiki Documentation

Not yet

# Changelog

Not yet

:cl: SuperSlayer, icons and sounds by Goonstation
rscadd: Flockmind midround antagonist
imageadd: Icons for various flock objects
imageadd: Icons for flock mobs
soundadd: Sounds for some flock thingies
tweak: Refactored CanAttack() hostile mob proc.
/:cl:

### CODE REVIEW

**Comment by ghost** [comment]
Me when github is a good site

**Comment by neeshacark** [comment]
so swarmers but more scary?

**Comment by ghost** [comment]
> so swarmers but more scary?

Yes. More scary swarmers with RTS elements.

**Comment by ghost** [comment]
It is ready, but needs someone to check the code because I am a dumbfuck and could fuck up with something.

File: code/__DEFINES/role_preferences.dm [hunk 1]
```diff
@@ -94,8 +96,12 @@ GLOBAL_LIST_INIT(special_roles, list(
 	ROLE_FUGITIVE,
 	ROLE_BLOODSUCKER = /datum/game_mode/bloodsucker,
 	ROLE_MONSTERHUNTER,
+	ROLE_SINFULDEMON,
 	ROLE_SPACE_DRAGON,
-	ROLE_SINFULDEMON
+	ROLE_SINFULDEMON,
+	ROLE_SINFULDEMON,
```
Comment (tattax) [inline code/__DEFINES/role_preferences.dm]: what

File: yogstation/code/_onclick/hud/flocktrace.dm [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/mob/camera/flocktrace/ClickOn(var/atom/A, var/params)
+	if(stored_action)
+		if(stored_action.perform_action(A))
+			qdel(stored_action)
+			stored_action = null
+			return
+	var/list/modifiers = params2list(params)
+	A.attack_flocktrace(src, modifiers)
+
+/atom/proc/attack_flocktrace(/mob/camera/flocktrace/user, var/list/modifiers)
```
Comment (tattax) [inline yogstation/code/_onclick/hud/flocktrace.dm]:
```suggestion
/mob/camera/flocktrace/ClickOn(atom/A, params)
	if(stored_action?.perform_action(A))
		qdel(stored_action)
		stored_action = null
		return
	var/list/modifiers = params2list(params)
	A.attack_flocktrace(src, modifiers)

/atom/proc/attack_flocktrace(mob/camera/flocktrace/user, list/modifiers)
```

File: yogstation/code/modules/antagonists/flockmind/abilities.dm [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+			if(istype(FT.stored_action, action.type))
+				qdel(FT.stored_action)
+				FT.stored_action = null
+				to_chat(owner, span_warning("You cancell your current action."))
+				return
+			qdel(FT.stored_action)
+			FT.stored_action = null			
+		FT.stored_action = new action(FT)
+	else
+		return
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/abilities.dm]: early return instead

File: yogstation/code/modules/antagonists/flockmind/abilities.dm [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+/datum/action/cooldown/flock/IsAvailable()
+	return (next_use_time <= world.time) && (isflockdrone(owner) || isflocktrace(owner))
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/abilities.dm:21]: cooldown define

File: yogstation/code/modules/antagonists/flockmind/abilities.dm [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+			FD.Goto(pick(surrounding_turfs), FD.move_to_delay)
+
+/datum/action/cooldown/flock/eject
+	name = "Eject"
+	desc = "Leave your current vessel."
+	button_icon_state = "eject"
+
+/datum/action/cooldown/flock/eject/Trigger()
+	if(!isflockdrone(owner))
+		owner.balloon_alert(owner, "Not in a flockdrone")
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/abilities.dm:60]: balloon alert enjoyers

File: yogstation/code/modules/antagonists/flockmind/abilities.dm [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+	if(length(targets))
+		playsound(owner, 'sound/misc/flockmind/flockmind_cast.ogg', 80, 1)
+		to_chat(owner, span_notice("You force open all the airlocks around you."))
+		StartCooldown()
+		sleep(1.5 SECONDS)
+		for(var/obj/machinery/door/airlock/A in targets)
+			A.open()
+	else
+		owner.balloon_alert(owner, "No valid targets")
+		return
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/abilities.dm]: switch this around and remove the else (!targets.len)

Comment (tattax) [reply yogstation/code/modules/antagonists/flockmind/abilities.dm]: also i'm pretty sure balloon alerts (on tg atleast) use full on lowercase

File: yogstation/code/modules/antagonists/flockmind/abilities.dm [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+/datum/action/cooldown/flock/radio_stun/Trigger()
+	var/list/targets = list()
+	for(var/mob/living/L in range(10, owner))
+		if(HAS_TRAIT(L, TRAIT_DEAF))
+			continue
+		if(L.stat == DEAD)
+			continue
+		var/obj/item/radio/headset/H = L.get_item_by_slot(ITEM_SLOT_EARS)
+		if(H && H.on && H.listening && istype(H))
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/abilities.dm]:
```suggestion
		if(H?.on && H.listening && istype(H))
```

File: yogstation/code/modules/antagonists/flockmind/abilities.dm [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+			targets += L
+	if(length(targets))
+		playsound(owner, 'sound/misc/flockmind/flockmind_cast.ogg', 80, 1)
+		to_chat(owner, span_notice("You transmit the worst static you can weave into the headsets around you."))
+		StartCooldown()
+		for(var/mob/living/L in targets)
+			var/distance = max(0,get_dist(get_turf(owner), get_turf(L)))
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/abilities.dm]:
```suggestion
			var/distance = max(0 ,get_dist(get_turf(owner), get_turf(L)))
```

File: yogstation/code/modules/antagonists/flockmind/abilities.dm [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+			if(L.soundbang_act(1, 20/max(1,distance), rand(1, 5)))
+				L.Paralyze(max(150/max(1,distance), 60))
+				to_chat(L, span_userdanger("Horrifying static bursts into your headset, disorienting you severely!"))
+				playsound(L, pick(list('sound/effects/radio_sweep1.ogg','sound/effects/radio_sweep2.ogg','sound/effects/radio_sweep3.ogg','sound/effects/radio_sweep4.ogg','sound/effects/radio_sweep5.ogg')), 100, 1)
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/abilities.dm]:
```suggestion
				playsound(L, pick(list('sound/effects/radio_sweep1.ogg','sound/effects/radio_sweep2.ogg','sound/effects/radio_sweep3.ogg','sound/effects/radio_sweep4.ogg','sound/effects/radio_sweep5.ogg')), 100, TRUE)
```

File: yogstation/code/modules/antagonists/flockmind/commands.dm [hunk 10]
```diff
@@ -0,0 +1,113 @@
+/datum/flock_command
+	var/mob/camera/flocktrace/daddy
+	var/atom/dude
+	var/messg = "Uhh you will do nothing cry about it"
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/commands.dm:4]: uhhhhhhh

File: yogstation/code/modules/antagonists/flockmind/commands.dm [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+	daddy = ded
+	if(daddy.stored_action && daddy.stored_action != src)
+		qdel(daddy.stored_action)
+		daddy.stored_action = src
+	if(A)
+		dude = A
+	if(PA)
+		parent_action = PA
+	if(messg)
+		to_chat(daddy, span_boldnotice(messg))
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/commands.dm:21]: probably you'd want better var names

File: yogstation/code/modules/antagonists/flockmind/commands.dm [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+		return FALSE
+	if(!FT.getFireLoss() && !FT.getBruteLoss())
+		FT.balloon_alert(daddy, "Already at full health")
+		return TRUE
+	if(FT.stat == DEAD)
+		FT.balloon_alert(daddy, "Dead, aborting")
+		return TRUE
+	FT.heal_ordered_damage(30, list(BRUTE, BURN))
+	FT.balloon_alert(daddy, "Sucessfully healed")
+	playsound(daddy, 'sound/misc/flockmind/flockmind_cast.ogg', 80, 1)
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/commands.dm]:
```suggestion
	playsound(daddy, 'sound/misc/flockmind/flockmind_cast.ogg', 80, TRUE)
```

File: yogstation/code/modules/antagonists/flockmind/commands.dm [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+	return TRUE
+
+/datum/flock_command/talk
+	messg = "Next living being with a working listening headset will allow you to transmit something to it. Also works on radio devices."
+
+/datum/flock_command/talk/perform_action(atom/A)
+	if(isliving(A))
+		var/mob/living/L = A 
+		var/obj/item/radio/headset/H = L.get_item_by_slot(ITEM_SLOT_EARS)
+		if(!H || !H.on || !H.listening || !istype(H))
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/commands.dm]:
```suggestion
		if(!H?.on || !H.listening || !istype(H))
```

File: yogstation/code/modules/antagonists/flockmind/commands.dm [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+			return TRUE
+		var/msg = input(daddy,"What would you like to transmit to [L]?","Narrowbeam Transmission","") as null|text
+		if(!msg)
+			return TRUE
+		to_chat(L, span_swarmer("You hear a crackling voice through your headset: [msg]"))
+		playsound(L, 'yogstation/sound/effects/radio1.ogg', 50, 1)
+		daddy.log_talk(msg, LOG_SAY)
+		to_chat(daddy, span_notice("You transmit a message to [L]: [msg]"))
+		return TRUE
+	else if(istype(A, /obj/item/radio))
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/commands.dm:96]: you can just if here specially when you returned on the last proc

File: yogstation/code/modules/antagonists/flockmind/commands.dm [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+		msg = trim(copytext(sanitize(msg), 1, MAX_MESSAGE_LEN))
+		var/name = daddy.name
+		daddy.name = "Unknown"
+		R.talk_into(daddy, msg, 0, "Unknown")
+		daddy.name = name
+		to_chat(daddy, span_notice("You transmit a message through [R]: [msg]"))
+		return TRUE
+	else 
+		to_chat(daddy, span_warning("[A] isn't a valid target!"))
+		return FALSE
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/commands.dm:113]: no else

File: yogstation/code/modules/antagonists/flockmind/compute.dm [hunk 16]
```diff
@@ -0,0 +1,29 @@
+/datum/component/flock_compute
+	var/compute_amount = 0
+	var/requires_alive = TRUE
+
+/datum/component/flock_compute/Initialize(initial_compute=0, requires_living_if_mob = TRUE)
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/compute.dm:5]: . = ..()

File: yogstation/code/modules/antagonists/flockmind/flock.dm [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+				to_chat(M, "[link] [rendered]")
+			else
+				to_chat(M, "[rendered]")
+
+/proc/get_flock_team(list/members)
+	if(GLOB.flock)
+		return GLOB.flock
+	else
+		var/datum/team/flock/F = new(members)
+		return F
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/flock.dm:45]: no else

File: yogstation/code/modules/antagonists/flockmind/flock.dm [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+		if(F.hud_used && istype(F.hud_used, /datum/hud/flocktrace))
+			var/datum/hud/flocktrace/flockhud = F.hud_used
+			flockhud.compute.update_counter(get_compute(TRUE))
+	if(acting)
+		var/num_drones =0
+		for(var/mob/M in GLOB.player_list)
+			if(isflockdrone(M) && M.stat != DEAD)
+				num_drones++
+		var/compute = get_compute(TRUE)
+		if(num_drones <= 0 && compute <= 0)
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/flock.dm:90]: !

File: yogstation/code/modules/antagonists/flockmind/flock_act.dm [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+		to_chat(drone, span_warning("You attempt to disintegrate [src] into resources."))
+		if(!do_mob(drone, src, 4 SECONDS))
+			return
+		if(QDELETED(src))
+			return
+		to_chat(drone, span_warning("You sucessfully transform [src] into your resources."))
+		drone.change_resources(maxHealth/2)
+		gib()
+	else
+		return
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/flock_act.dm]: early return

File: yogstation/code/modules/antagonists/flockmind/flock_act.dm [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+		to_chat(src, span_warning("There is already a barricade here."))
+		return
+	new /obj/structure/grille/flock (src)
+	drone.change_resources(-25)
+
+/obj/structure/grille/flock/flock_act(mob/living/simple_animal/hostile/flockdrone/drone)
+	if(broken)
+		to_chat(drone, span_warning("[src] Is to damaged to extract any resources!"))
+	to_chat(drone, span_notice("You begin deconstructing [src]..."))
+	if(!do_after(drone, src, 3 SECONDS))
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/flock_act.dm]: wrong arg order

File: yogstation/code/modules/antagonists/flockmind/flock_act.dm [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+		new /obj/structure/grille/flock (loc)
+	else
+		new /obj/structure/grille/flock/broken (loc)
+	qdel(src)
+
+/turf/closed/wall/feather/flock_act(mob/living/simple_animal/hostile/flockdrone/drone)
+	if(!drone)
+		return
+	to_chat(drone, span_notice("You begin deconstructing [src]..."))
+	if(!do_after(drone, src, 5 SECONDS))
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/flock_act.dm]: .

File: yogstation/code/modules/antagonists/flockmind/flock_act.dm [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+		return
+	if(QDELETED(src) || !istype(/turf/closed/wall/feather, src) || !src)
+		return
+	to_chat(drone, span_notice("You deconstruct [src]."))
+	TerraformTurf(/turf/open/floor/feather)
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/flock_act.dm:130]: at this point why not have a openturf feather define

File: yogstation/code/modules/antagonists/flockmind/flockdrone.dm [hunk 23]
```diff
@@ -0,0 +1,378 @@
...
+/mob/living/simple_animal/hostile/flockdrone/OpenFire(atom/A)
+	if(!ckey)
+		handle_AI_intent_change(target)
+	. = ..()
+
+/mob/living/simple_animal/hostile/flockdrone/Shoot(atom/targeted_atom)
+	if(a_intent == INTENT_HELP)
+		projectiletype = /obj/item/projectile/beam/disabler/flock
+	else 
+		projectiletype = /obj/item/projectile/beam/flock
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/flockdrone.dm]: could probably do this in one line but this works

File: yogstation/code/modules/antagonists/flockmind/structures/cage.dm [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+	can_be_unanchored = FALSE
+	canSmoothWith = null
+	flags_1 = NODECONSTRUCT_1
+	var/processing = FALSE
+	var/resources = 0
+
+/obj/structure/destructible/flock/cage/user_unbuckle_mob(mob/living/buckled_mob, mob/living/user)
+	if(has_buckled_mobs())
+		for(var/buck in buckled_mobs) //breaking a nest releases all the buckled mobs, because the nest isn't holding them down anymore
+			var/mob/living/M = buck
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/structures/cage.dm]: early return + full var in the loop rather then setting it afterwards

File: yogstation/code/modules/antagonists/flockmind/structures/cage.dm [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+	unbuckle_all_mobs()
+
+/obj/structure/destructible/flock/cage/Destroy()
+	. = ..()
+	if(processing)
+		STOP_PROCESSING(SSobj, src)
+
+/obj/structure/destructible/flock/cage/process()
+	var/valid_dudes = 0
+	if(has_buckled_mobs())
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/structures/cage.dm]: probably better var name + ditto

File: yogstation/code/modules/antagonists/flockmind/structures/cage.dm [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+		for(var/buck in buckled_mobs)
+			var/mob/living/carbon/C = buck
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/structures/cage.dm]: .

File: yogstation/code/modules/antagonists/flockmind/structures/relay.dm [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+	if(world.time >= charge_time_length/2) // halfway point, start doing more
+		if(icon_state == "structure-relay")
+			icon_state = "structure-relay-glow"
+
+		for(var/mob/M in GLOB.player_list)
+			if(prob(20))
+				playsound(M, pick(list('sound/effects/radio_sweep1.ogg','sound/effects/radio_sweep2.ogg','sound/effects/radio_sweep3.ogg','sound/effects/radio_sweep4.ogg','sound/effects/radio_sweep5.ogg'), 50, 1))
+				if(prob(50))
+					to_chat(M, span_italics("the signal will set you free"))
+	if(world.time >= charge_time_length)
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/structures/relay.dm]: cooldown define

File: yogstation/code/modules/antagonists/flockmind/structures/sentinel.dm [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+	var/datum/team/flock/flock = get_flock_team()
+	if(charge < 100)
+		if(flock.get_compute() < online_compute_cost) //We can't afford to charge if no compute(
+			return 
+		if(compute_provided != online_compute_cost)
+			change_compute_amount(online_compute_cost)
+			icon_state = "sentinel"
+		charge += charge_per_tick
+		return
+	if(compute_provided != 0)
```
Comment (tattax) [inline yogstation/code/modules/antagonists/flockmind/structures/sentinel.dm]:
```suggestion
	if(compute_provided)
```

**Comment by Bibby0110** [comment]
you have to put the goon sprites in the correct folder, this is very iffy licensing wise

**Comment by ghost** [comment]
> you have to put the goon sprites in the correct folder, this is very iffy licensing wise

Okay, i will do that. Also if you are going to do something like testmerging it, wait on, because it is very broken, currently(tested with friends)

File: code/__DEFINES/role_preferences.dm (hunk 1 above)
Comment (ghost) [reply code/__DEFINES/role_preferences.dm]: Oops

File: yogstation/code/modules/antagonists/flockmind/commands.dm (hunk 10 above)
Comment (ghost) [reply yogstation/code/modules/antagonists/flockmind/commands.dm:4]: It is a unacessable ingame

**Comment by ghost** [comment]
Notes for me:
1. Intent HUD button for some reason is segmented, while it shouldn't be. 
2. Intents do not work
3. Movement mode button doesn't appear
4. Some flockmind abilities dont work
5. Flocktraces have weird movement and vision (maybe fixed)
6. Flockdrones are not able to convert objects and their AI is completely broken (maybe fixed)
7. Sentinel tesla zaps dont deal damage and work really strange (maybe fixed)

File: yogstation/code/modules/antagonists/flockmind/abilities.dm (hunk 3 above)
Comment (ghost) [reply yogstation/code/modules/antagonists/flockmind/abilities.dm]: Done

File: yogstation/code/modules/antagonists/flockmind/abilities.dm (hunk 6 above)
Comment (ghost) [reply yogstation/code/modules/antagonists/flockmind/abilities.dm]: Done. Also no, on TG balloon alerts are sometimes are not fully on lowercase.

File: yogstation/code/modules/antagonists/flockmind/structures/cage.dm (hunk 24 above)
Comment (ghost) [reply yogstation/code/modules/antagonists/flockmind/structures/cage.dm]: Done

File: yogstation/code/modules/antagonists/flockmind/flock_act.dm (hunk 21 above)
Comment (ghost) [reply yogstation/code/modules/antagonists/flockmind/flock_act.dm]: Done

**Comment by ghost** [comment]
Maybe it is ready for being testmerged, but I am not sure

**Comment by JamieD1** [comment]
> Maybe it is ready for being testmerged, but I am not sure

Can you test it a bit more,

**Comment by ghost** [comment]
> Can you test it a bit more,

Sure

**Comment by ghost** [comment]
Aaaa fuck

**Comment by SuperSlayer0** [comment]
Cool, thanks github for throwing my 2000 lines of code project into trash

**Comment by ghost** [comment]
Idk what to do with this thing, since I am unable to test it myself

**Pull request closed without merging**